src/generic/linalg.jl

```diff
@@ -0,0 +1,478 @@
+
+        Base.LinAlg.At_mul_B!(C::AbstractMatrix, A::$Typ1, B::$Typ2) = BandedMatrices.banded_matmatmul!(C, 'T', 'N', A, B)
+        Base.LinAlg.A_mul_Bt!(C::AbstractMatrix, A::$Typ1, B::$Typ2) = BandedMatrices.banded_matmatmul!(C, 'N', 'T', A, B)
+        Base.LinAlg.At_mul_Bt!(C::AbstractMatrix, A::$Typ1, B::$Typ2) = BandedMatrices.banded_matmatmul!(C, 'T', 'T', A, B)
+
+        # override the Ac_mul_B, A_mul_Bc and Ac_mul_c for real values
+        Base.LinAlg.Ac_mul_B!(C::AbstractMatrix{T}, A::$Typ1{U}, B::$Typ2{V}) where {T, U<:Real, V} = BandedMatrices.banded_matmatmul!(C, 'T', 'N', A, B)
+        Base.LinAlg.A_mul_Bc!(C::AbstractMatrix{T}, A::$Typ1{U}, B::$Typ2{V}) where {T, U, V<:Real} = BandedMatrices.banded_matmatmul!(C, 'N', 'T', A, B)
+        Base.LinAlg.Ac_mul_Bc!(C::AbstractMatrix{T}, A::$Typ1{U}, B::$Typ2{V}) where {T, U<:Real, V<:Real} = BandedMatrices.banded_matmatmul!(C, 'T', 'T', A, B)
+
+        function Base.:*(A::$Typ1{T}, B::$Typ2{V}) where {T, V}
+            n, m = size(A,1), size(B,2)
+            Y = BandedMatrix(promote_type(T,V), n, m, prodbandwidths(A, B)...)
+            A_mul_B!(Y, A, B)
+        end
+
+        Base.LinAlg.Ac_mul_B(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = Ac_mul_B!(BandedMatrices.banded_similar('C', 'N', A, B, promote_type(T, V)), A, B)
+        Base.LinAlg.A_mul_Bc(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = A_mul_Bc!(BandedMatrices.banded_similar('N', 'C', A, B, promote_type(T, V)), A, B)
+        Base.LinAlg.Ac_mul_Bc(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = Ac_mul_Bc!(BandedMatrices.banded_similar('C', 'C', A, B, promote_type(T, V)), A, B)
+
+        Base.LinAlg.At_mul_B(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = At_mul_B!(BandedMatrices.banded_similar('T', 'N', A, B, promote_type(T, V)), A, B)
+        Base.LinAlg.A_mul_Bt(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = A_mul_Bt!(BandedMatrices.banded_similar('N', 'T', A, B, promote_type(T, V)), A, B)
+        Base.LinAlg.At_mul_Bt(A::$Typ1{T}, B::$Typ2{V}) where {T, V} = At_mul_B!(BandedMatrices.banded_similar('T', 'T', A, B, promote_type(T, V)), A, B)
+    end
+    esc(ret)
+end
+
+macro banded_banded_linalg(Typ1, Typ2)
+    ret = quote
+        BandedMatrices.@_banded_banded_linalg($Typ1, $Typ2)
+    end
+    if Typ1 ≠ Typ2
+        ret = quote
+            $ret
+            BandedMatrices.@_banded_banded_linalg($Typ2, $Typ1)
+            BandedMatrices.@banded_banded_banded_linalg($Typ1, $Typ1, $Typ2)
+            BandedMatrices.@banded_banded_banded_linalg($Typ2, $Typ2, $Typ1)
+        end
+    end
+    esc(ret)
+end
+
+# add banded linear algebra routines for Typ implementing the BandedMatrix interface
+macro banded_linalg(Typ)
+    ret = quote
+        BandedMatrices.@banded_banded_linalg($Typ, $Typ)
+        # default is to use dense axpy!
+        BandedMatrices.banded_axpy!(a::Number, X::$Typ, Y::AbstractMatrix) =
+                BandedMatrices.banded_dense_axpy!(a, X, Y)
```

Comment:
perhaps there is obvious, but why not using `banded_generic_axpy` if the interface has been implemented?